### PR TITLE
American option baron adesi whaley

### PR DESCRIPTION
**Pull request by esztermarton**
Implementing Baron-Adesi Whaley approximation of black scholes formula

### CODE REVIEW

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+
+  :param sigma:
+  :param x:
+  :param t:
+  :param r:
+  :param b:
+  :param s:
+  :param call:
+  :param dtype:
+  :return:
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: Please remove

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    is_call_options, am_indices) if is_call_options is not None else None
+
+  am_prices = _option_price(
+    sigma_am, x_am, t_am, spots=s_am, discount_rates=r_am,
+    cost_of_carries=b_am, is_call_options=is_call_am, dtype=dtype, name=name)
+
+  return tf.cond(
+    tf.reduce_all(calculate_eu),
+    lambda: eu_prices,
+    lambda: tf.where(calculate_eu, eu_prices, am_prices))
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: Here both `tensor`s eu_prices and tf.where(calculate_eu, eu_prices, am_prices) are computed (see documentation [here](https://www.tensorflow.org/api_docs/python/tf/cond))

Comment (cyrilchim) [reply tf_quant_finance/black_scholes/approximations/american_option.py]: If this is an intension, you can remove `tf.cond` (which I would recommend doing anyway)

File: tf_quant_finance/black_scholes/approximations/american_option.py (hunk 1 above)
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: please remove

File: tf_quant_finance/black_scholes/approximations/american_option_test.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+from absl.testing import parameterized
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+
+import tf_quant_finance as tff
+from tensorflow.python.framework import test_util  # pylint: disable=g-direct-tensorflow-import
+
+
+print(dir(tff))
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option_test.py]: please remove

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+  :param b:
+  :param s:
+  :param call:
+  :param dtype:
+  :return:
+  """
+  is_call_options = None if call else tf.constant(False)
+  sign = 1 if call else -1
+
+  M = 2 * r / sigma ** 2
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: all names should be decapitalized: 'm' insead of `M`. I assume, you are using notations from the paper. Please add a comment that this is indeed the case

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+    shape = tf.shape(spots, out_type=call_indices.dtype)
+
+    call_or_zero = tf.scatter_nd(call_indices, american_call_prices, shape)
+    put_or_zero = tf.scatter_nd(put_indices, american_put_prices, shape)
+    return tf.where(is_call_options, call_or_zero, put_or_zero)
+
+
+def _option_price_put_or_call(sigma, x, t, r, b, call, dtype):
+  """ Calculates q, a for the american option price formula.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: Not sure what q and a means. Please add nore informative description

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  q_inf = _calc_q(N, M, call=call)
+  s_inf = x / (1 - 1 / q_inf)
+  h = -(sign * b * t + 2 * sigma * tf.sqrt(t)) * sign * (x / (s_inf - x))
+  if call:
+    s_seed = x + (s_inf - x) * (1 - tf.exp(h))
+  else:
+    s_seed = s_inf + (x - s_inf) * tf.exp(h)
+
+  s_crit, _, _ = implied_vol_newton.newton_root_finder(
+    value_and_gradient, s_seed, dtype=dtype)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: You need to add `tolerance` and `max_iterations` parameters. Also, the output of american options pricer should include  convergence results. See the [doctring of implied_vol](https://github.com/google/tf-quant-finance/blob/514236ce735ee2d6bdd9fb9981ce79da4dcaa1cd/tf_quant_finance/black_scholes/implied_vol_newton_root.py#L23)

Comment (cyrilchim) [reply tf_quant_finance/black_scholes/approximations/american_option.py]: Also, please pass the arguments by name

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+    s_crit, x, sigma, b, t))) * s_crit / q_ - sign * (s_crit - x)
+
+  value_and_gradient = lambda price: tff.math.value_and_gradient(value, price)
+
+  # Calculate seed value for critical spot price
+  q_inf = _calc_q(N, M, call=call)
+  s_inf = x / (1 - 1 / q_inf)
+  h = -(sign * b * t + 2 * sigma * tf.sqrt(t)) * sign * (x / (s_inf - x))
+  if call:
+    s_seed = x + (s_inf - x) * (1 - tf.exp(h))
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py:271]: You don't need tf.gather_nd above. use tf.where here to build a starting point

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+  M = 2 * r / sigma ** 2
+  N = 2 * b / sigma ** 2
+  K = 1 - tf.exp(- r * t)
+  q_ = _calc_q(N, M, K, call)
+
+  value = lambda s_crit: vanilla_prices.option_price(
+    volatilities=sigma, strikes=x, expiries=t, spots=s_crit, discount_rates=r,
+    cost_of_carries=b, is_call_options=is_call_options, dtype=dtype) + sign * (
+      1 - tf.exp((b - r) * t) * _ncdf(sign * _calc_d1(
+    s_crit, x, sigma, b, t))) * s_crit / q_ - sign * (s_crit - x)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py:262]: You may build the target function with `tf.where` to eliminate tf.gather_nd above. Does it make sense?

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    american_call_prices = tf.where(
+      is_less_than_s_crit,
+      eu_call_prices_or_zero + a2 * (s_call / s_crit_call) ** q2,
+      s_call - x_call)
+
+    if is_call_options is None:
+      return american_call_prices
+
+    # Put option
+    x_put = tf.gather_nd(strikes, put_indices)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: Let's try to remove gather_nd.

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+    sigma_put = tf.gather_nd(volatilities, put_indices)
+    s_put = tf.gather_nd(spots, put_indices)
+    t_put = tf.gather_nd(expiries, put_indices)
+    r_put = tf.gather_nd(discount_rates, put_indices)
+    b_put = tf.gather_nd(cost_of_carries, put_indices)
+
+    q1, a1, s_crit_put = _option_price_put_or_call(
+      sigma_put, x_put, t_put, r_put, b_put, False, dtype)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: I think this function can be rewritten with tf.where without any need for gather_nd

**Review by cyrilchim — COMMENTED** [review]
This is good but I would rather remove `tf.gather_nd` with tf.where (let me know if you have any issues with that)

File: tf_quant_finance/black_scholes/approximations/american_option.py (hunk 2 above)
Comment (esztermarton) [reply tf_quant_finance/black_scholes/approximations/american_option.py]: The intention was that am_prices, while the graph can be constructed (which I thought is how cond works), wouldn't be calculated unless there are some that require it, and if all are eu, then no actual computation would happen there, only graph construction.

File: tf_quant_finance/black_scholes/approximations/american_option.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+def _calc_q(n, m, sign, k=1):
+  return ((1 - n) + sign * tf.sqrt((n - 1) ** 2 + 4 * m / k)) / 2
+
+
+def _ncdf(x):
+  return (tf.math.erf(x / _SQRT_2) + 1) / 2
+
+
+def _npdf(x):
+  return tf.exp(-0.5 * x ** 2) / tf.math.sqrt(2 * np.pi)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option.py]: tf.math.sqrt(2 * np.pi) is always of dtype tf.float64. Could you please define a global constant like _SQRT_2?

File: tf_quant_finance/black_scholes/approximations/american_option_test.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+import tensorflow.compat.v2 as tf
+
+import tf_quant_finance as tff
+from tensorflow.python.framework import test_util  # pylint: disable=g-direct-tensorflow-import
+
+
+@test_util.run_all_in_graph_and_eager_modes
+class AmericanPrice(parameterized.TestCase, tf.test.TestCase):
+  """Tests for methods for the american pricing module."""
+
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/approximations/american_option_test.py:29]: Could you please add a dtype as a parameter to at least one of your test?

**Review by cyrilchim — COMMENTED** [review]
Looks amazing! Thank you, Rebeka. Just two minor comments. Could you please push it to the repo once done?

**Pull request merged**